# Importing librariesand datasets 

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os


In [2]:
# creating a path string

path=r"C:\Users\Lemlem TH\Desktop\CareerFoundry\Course\Data Immersion\Achievement 4\Instacart Basket Analysis"


In [3]:
# Importing the ords_prods_merge_derived.pkl dataset as ords_prods_merge dataframe

ords_prods_merge=pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data','ords_prods_merge_derived.pkl' ))

In [4]:
 df=ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 19)

In [6]:
df.head()

order_id  user_id  order_number  orders_day_of_week  hour_at_time_of_order  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days busiest_periods_of_day  
0  Mid-range product  Regularly busy  Regularly days            Most orders  
1  Mid-range product  Regularly busy    Slowest days         Average orders  
2  Mid-range product  Regularly busy    Slowest days            Most orders  
3  Mid-range product      Least busy    Slowest days         Average orders  
4  Mid-range product      Least busy    Slowest days            Most orders

# Grouping Data with pandas

In [7]:
# Grouping the column product_name
df.groupby('product_name')

# Aggregating Data with agg()

In [8]:
# Performing a single aggregation

df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# Performing aggregation without the agg() function

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# Performing aggregation with dot notation (but not recommended for ease of understanding)

df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [11]:
# Performing Multiple Aggregations

df.groupby('department_id').agg({'order_number':['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Aggregating Data with transform()

In [12]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [13]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  \
0    2539329        1             1                   2   
1    2398795        1             2                   3   
2     473747        1             3                   3   
3    2254736        1             4                   4   
4     431534        1             5                   4   
5    3367565        1             6                   2   
6     550135        1             7                   1   
7    3108588        1             8                   1   
8    2295261        1             9                   1   
9    2550362        1            10                   4   
10   2968173       15            15                   1   
11   1870022       15            17                   2   
12   1911383       15            18                   2   
13   2715276       15            21                   1   
14    487368       15            22                   1   

    hour_at_time_of_order  days_since_prior_order  product_id  \
0                       8                     0.0         196   
1                       7                    15.0         196   
2                      12                    21.0         196   
3                       7                    29.0         196   
4                      15                    28.0         196   
5                       7                    19.0         196   
6                       9                    20.0         196   
7                      14                    14.0         196   
8                      16                     0.0         196   
9                       8                    30.0         196   
10                      9                     7.0         196   
11                     16                     8.0         196   
12                     11                     7.0         196   
13                      9                     7.0         196   
14                     10                    14.0         196   

    add_to_cart_order  reordered  Unnamed: 0 product_name  aisle_id  \
0                   1          0         195         Soda        77   
1                   1          1         195         Soda        77   
2                   1          1         195         Soda        77   
3                   1          1         195         Soda        77   
4                   1          1         195         Soda        77   
5                   1          1         195         Soda        77   
6                   1          1         195         Soda        77   
7                   2          1         195         Soda        77   
8                   4          1         195         Soda        77   
9                   1          1         195         Soda        77   
10                  2          0         195         Soda        77   
11                  6          1         195         Soda        77   
12                  1          1         195         Soda        77   
13                  1          1         195         Soda        77   
14                  1          1         195         Soda        77   

    department_id  prices _merge    price_range_loc     busiest_day  \
0               7     9.0   both  Mid-range product  Regularly busy   
1               7     9.0   both  Mid-range product  Regularly busy   
2               7     9.0   both  Mid-range product  Regularly busy   
3               7     9.0   both  Mid-range product      Least busy   
4               7     9.0   both  Mid-range product      Least busy   
5               7     9.0   both  Mid-range product  Regularly busy   
6               7     9.0   both  Mid-range product  Regularly busy   
7               7     9.0   both  Mid-range product  Regularly busy   
8               7     9.0   both  Mid-range product  Regularly busy   
9               7     9.0   both  Mid-range product      Least busy   
10              7     9.0   both  Mid-range product  Regularly busy   
11              7     9.0   b

# Deriving Columns with loc()

In [14]:
# Creating a flag column that assigns 'loyalty' label

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'


In [15]:
# Print the frequency of the newly created flag column 

ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

# Tasks

In [16]:
# Step 2 find the aggregated mean of the 'order_number' column grouped by 'departemnt_id'

ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Step 3
The full dataset contains 21 departments whereas the subset contained only 8 departments. The highest value of mean order_number is 22.9 ( for department 21) in the full dataset. Whereas the highest mean value is 196 for department 16 for the sub-set dataset

In [17]:
#.  Step 4 create a loyalty flag using the transform() and loc() functions. First aggregate data with transform () function

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [18]:
# Assign flag values using conditions 

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_at_time_of_order  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0  ... aisle_id  department_id  prices  _merge    price_range_loc  \
0         195  ...       77              7     9.0    both  Mid-range product   
1         195  ...       77              7     9.0    both  Mid-range product   
2         195  ...       77              7     9.0    both  Mid-range product   
3         195  ...       77              7     9.0    both  Mid-range product   
4         195  ...       77              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_periods_of_day max_order  \
0  Regularly busy  Regularly days            Most orders        10   
1  Regularly busy    Slowest days         Average orders        10   
2  Regularly busy    Slowest days            Most orders        10   
3      Least busy    Slowest days         Average orders        10   
4      Least busy    Slowest days            Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [20]:
# Print the frequency of the loyalty flag column

ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [21]:
# Step 5 Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers 

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

On average regular and new customers spend more than loyal customers.

In [22]:
# Step 6 Create a spending flag. First aggregate data with the transform() function 

ords_prods_merge['avg_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [23]:
# Assign flag values using conditions 

ords_prods_merge.loc[ords_prods_merge['avg_spending'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_spending'] >= 10, 'spending_flag'] = 'High spender'

In [24]:
# Print the frequency of the spending flag column

ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [25]:
# Step 7 Create an order frequency flag. First aggregate data with the transform functin

ords_prods_merge['ord_freq'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
# Assign flag values using conditions 

ords_prods_merge.loc[ords_prods_merge['ord_freq'] > 20, 'order_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['ord_freq'] <= 20) & (ords_prods_merge['ord_freq'] > 10), 'order_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['ord_freq'] <= 10, 'order_flag'] = 'Frequent customer'

In [27]:
# Print the frequency of the spending flag column

ords_prods_merge['order_flag'].value_counts(dropna=False)

order_flag
Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: count, dtype: int64

In [29]:
# Exporting the dataframe to pkl as ords_prods_merge_derived_agg.pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived_agg.pkl'))